### PR TITLE
mesonbuild/mtest: plumb and report a test's start time

### PR DESCRIPTION
**Pull request by camilo-celis**
I have had cases where having the start time reported on the stdout of a test could be useful (especially when tests are run in parallel). This patch simply plumbs that data to TestRun and uses it to include it on the `get_log` string representation.

### CODE REVIEW

File: mesonbuild/mtest.py [hunk 1]
```diff
@@ -331,11 +331,12 @@ def make_exitcode(cls, test: 'TestSerialisation', test_env: typing.Dict[str, str
             res = TestResult.EXPECTEDFAIL if bool(returncode) else TestResult.UNEXPECTEDPASS
         else:
             res = TestResult.FAIL if bool(returncode) else TestResult.OK
-        return cls(test, test_env, res, returncode, duration, stdo, stde, cmd)
+        return cls(test, test_env, res, returncode, starttime, duration, stdo, stde, cmd)
 
     @classmethod
     def make_tap(cls, test: 'TestSerialisation', test_env: typing.Dict[str, str],
-                 returncode: int, duration: float, stdo: str, stde: str,
+                 returncode: int, starttime:float, duration: float,
```
Comment (ignatenkobrain) [inline mesonbuild/mtest.py]: __[Flake8]__ missing whitespace after ':' ([see on Sider](https://sider.review/gh/repos/19784232/pulls/6142#issue-4807815))

<sub>posted by Sider</sub>

Comment (camilo-celis) [reply mesonbuild/mtest.py]: @ignatenkobrain Yes, just saw that and pushed a commit seconds ago. Thank you for the heads-up.

**Comment by scivision** [comment]
I would like this feature. Handy for if some hardware or external system were involved, could help trace some difficult problems.

**Comment by camilo-celis** [comment]
> I would like this feature. Handy for if some hardware or external system were involved, could help trace some difficult problems.

That's exactly the use case I was trying to use meson to help me with :) correlating tests running and some sort of metrics where only time stamps are available.

**Comment by scivision** [comment]
```
concurrent.futures.process._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/usr/lib/python3.7/concurrent/futures/process.py", line 239, in _process_worker
    r = call_item.fn(*call_item.args, **call_item.kwargs)
  File "/root/run_project_tests.py", line 324, in run_test
    return _run_test(testdir, build_dir, install_dir, extra_args, compiler, backend, flags, commands, should_fail)
  File "/root/run_project_tests.py", line 409, in _run_test
    (returncode, tstdo, tstde, test_log) = run_test_inprocess(test_build_dir)
  File "/root/run_project_tests.py", line 283, in run_test_inprocess
    returncode_test = mtest.run_with_args(['--no-rebuild'])
  File "/root/mesonbuild/mtest.py", line 1059, in run_with_args
    return run(options)
  File "/root/mesonbuild/mtest.py", line 1045, in run
    return th.doit()
  File "/root/mesonbuild/mtest.py", line 802, in doit
    self.run_tests(tests)
  File "/root/mesonbuild/mtest.py", line 954, in run_tests
    self.drain_futures(futures)
  File "/root/mesonbuild/mtest.py", line 970, in drain_futures
    self.process_test_result(result.result())
  File "/usr/lib/python3.7/concurrent/futures/_base.py", line 428, in result
    return self.__get_result()
  File "/usr/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/usr/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/root/mesonbuild/mtest.py", line 489, in run
    return TestRun(self.test, self.test_env, TestResult.SKIP, GNU_SKIP_RETURNCODE, 0.0, skip_stdout, None, None)
TypeError: __init__() missing 1 required positional argument: 'cmd'
"""
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "run_cross_test.py", line 58, in <module>
    sys.exit(main())
  File "run_cross_test.py", line 55, in main
    return runtests(options.cross_file, options.failfast)
  File "run_cross_test.py", line 37, in runtests
    run_tests(commontests, 'meson-cross-test-run', failfast, ['--cross-file', cross_file])
  File "/root/run_project_tests.py", line 659, in run_tests
    return _run_tests(all_tests, log_name_base, failfast, extra_args)
  File "/root/run_project_tests.py", line 716, in _run_tests
    result = result.result()
  File "/usr/lib/python3.7/concurrent/futures/_base.py", line 435, in result
    return self.__get_result()
  File "/usr/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
TypeError: __init__() missing 1 required positional argument: 'cmd'
```

**Comment by camilo-celis** [comment]
> ```
> concurrent.futures.process._RemoteTraceback: 
> """
. . .
>   File "/usr/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
>     raise self._exception
> TypeError: __init__() missing 1 required positional argument: 'cmd'
> ```

I just saw that too. Hopefully after https://github.com/mesonbuild/meson/pull/6142/commits/dca7946159ea9b2bf217bff4367509250680936d all test pass.

**Comment by jpakkane** [comment]
Thanks.

**Pull request merged**